[10.4] Add section on speech recognition (#294)
huggingface/course: chapters/en/chapter10/4.mdx

# Automatically recognising speech

In this section, we’ll take a look at how Transformers can be used to convert spoken speech into text, a task known _speech recognition_.

TODO: insert Speech -> Model -> Text diagram

Speech recognition, also known as automatic speech recognition (ASR) or speech-to-text (STT), is one of the most popular and exciting spoken language processing tasks. It’s used in a wide range of applications, including dictation, voice assistants, video captioning and meeting transcriptions.

You’ve probably made use of a speech recognition system many times before without realising! Consider the digital assistant in your smartphone device (Siri, Google Assistant, Alexa). When you use these assistants, the first thing that they do is transcribe your spoken speech to written text, ready to be used for any downstream tasks.

Image: https://images.app.goo.gl/J6SU6RYryXXVT9uj7

> Demo: https://huggingface.co/facebook/wav2vec2-base-960h

Reviewer: For the demo, I suggest embedding a Gradio Space directly in the page, similar to how we did it in the NLP tasks: https://moon-ci-docs.huggingface.co/course/pr_294/en/chapter7/2?fw=pt

Author: (Attempted) resolve in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

Speech recognition is challenging as it requires joint knowledge of audio and text. The input audio might have lots of background noise, making it difficult to pick-out the spoken speech. The written text might have characters which don’t have an acoustic sound, such as punctuation, which are difficult to infer from audio alone. These are all hurdles we’ll have to tackle when building effective speech recognition systems.

> [//]: # Speech recognition is challenging as it requires a model to leverage knowledge about two domains concurrently: the input audio and the output text. Systems need to extract the relevant features from the input audio, possibly disentangling the spoken speech from background noise. From these features, they must then infer the contextual relation of words and sentences to form accurate and coherent sentences.

Reviewer: I prefer the text above, so suggest deleting this paragraph

Author: Resolved in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

Before we pick a dataset, we first need to understand some of the key defining features.

#### 1. Number of hours
Simply put, the number of training hours indicates how large the dataset is. It’s analogous to the number of training examples in an NLP dataset. However, bigger datasets aren’t necessarily better. If we want a model that generalises well, we want a diverse dataset with lots of different speakers, domains and speaking styles.

#### 2. Domain
The domain entails where the data was sourced from, whether it be audiobooks, podcasts, YouTube or financial meetings. Each domain has a different distribution of data. For example, audiobooks are recorded in high-quality studio conditions (with no background noise) and text that is taken from written literature. Whereas for YouTube, the audio likely contains more background noise and a more informal style of writing.

We need to match our domain to the conditions we anticipate at inference time. For instance, if we train our model on audiobooks, we can’t expect our model to perform well in noisy environments.

#### 3. Speaking style
The speaking style falls into one of two categories:

> * Narrated: read from a script

Reviewer: For some reason, this list doesn't render correctly on the frontend - maybe try adding a new line?
```suggestion

* Narrated: read from a script
```

Author: Resolved in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

* Spontaneous: un-scripted, conversational speech

The audio and text data reflect the style of speaking. Since narrated text is scripted, it tends to be spoken articulately and without any errors:

“Consider the task of training a model on a speech recognition dataset”

Whereas for spontaneous speech, we can expect a more colloquial style of speech, with the inclusion of repetitions, hesitations and false-starts:

“Let’s <uhm> take a look at how <silence> you’d you’d go about training a model on <uhm> a sp- speech recognition dataset” (this is an extreme example)

#### 4. Transcription style

> Whether the text has punctuation, casing or both. If we want a system to generate fully formatted text that could be used for a publication or meeting transcription, we require training data with punctuation and casing. If we just require the spoken words in an unformatted structure, neither punctuation nor casing are necessary.

Reviewer: ```suggestion
Transcription style refers to whether the text has punctuation, casing or both. If we want a system to generate fully formatted text that could be used for a publication or meeting transcription, we require training data with punctuation and casing. If we just require the spoken words in an unformatted structure, neither punctuation nor casing are necessary.
```

Author: Resolved in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

> A summary of datasets on the HF Hub:

Reviewer: ```suggestion
Here is a summary of popular speech recognition datasets on the Hugging Face Hub:
```

Author: Resolved in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

|                | Train Hours | Domain                                | Speaking Style         | Transcription Casing | Transcription Punctuation | Recommended Usage                                 |
|----------------|-------------|---------------------------------------|------------------------|----------------------|---------------------------|---------------------------------------------------|

> | LibriSpeech    | 960         | Audiobooks                            | Narrated               | ❌                    | ❌                         | Academic benchmarks                               |

Reviewer: I love this table - super handy for newcomers!! 

I suggest using the dataset ID so that it's easier to search on the Hub (or wrap with a URL that directs to the dataset page):

```suggestion
| `librispeech_asr`    | 960         | Audiobooks                            | Narrated               | ❌                    | ❌                         | Academic benchmarks                               |
```

Author: Great point! Have added wrapped URL links to each individual dataset's page on the Hub: https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

| TEDLIUM        | 452         | TED talks                             | Narrated               | ❌                    | ❌                         | Technical scientific, political and social topics |
| Voxpopuli      | 543         | European Parliament recordings        | Spontaneous            | ❌                    | ❌                         | Non-native English speakers                       |
| GigaSpeech     | 10000       | Audiobook, podcast, youtube           | Narrated & spontaneous | ❌                    | ✅                         | Robustness over multiple domains                  |
| SPGISpeech     | 5000        | Financial meetings                    | Narrated & spontaneous | ✅                    | ✅                         | Fully formatted transcriptions                    |

Let’s take an example where we want to train a speech recognition system to transcribe speech on the topic of machine learning (highly relevant!). Our speech is going to be narrated (scripted), and we’re not worried about punctuation or casing on our text output. From our reference table, it looks like TED-LIUM is a good choice of dataset!

> [//]: # On review, this sub-section might be too involved. Happy to omit it and dive straight into how to load a dataset.

Reviewer: Personally, I think the content is excellent and informative - we are not following the same format as the `transformers` tutorials (which tend to be much more to the point / focus on the code)

Author: Okay brill! Retaining it in that case 🙂

### How to load an audio dataset

> Before we can handle any audio datasets, we need to make sure we have the right dependencies installed. Specifically, we'll need the "Audio" feature from datasets: https://huggingface.co/docs/datasets/installation#audio. This should take care of the main Python packages we need to read audio files from byte form and convert them to arrays.

Reviewer: We usually try to avoid having explicit URLs in the course and prefer having explicit code snippets for key steps like installation. In this case, you could have something like:

```suggestion
Before we can handle any audio datasets, we need to make sure we have the right dependencies installed. Specifically, we'll need the [`audio` feature](https://huggingface.co/docs/datasets/installation#audio) from 🤗 Datasets: 

```
pip install datasets[audio]
```

This should take care of the main Python packages we need to read audio files from byte form and convert them to arrays.
```

Author: Thanks for the heads up! Have followed your suggestion in https://github.com/huggingface/course/pull/294/commits/9d111104b597c83c0eb05f25f36544eed01eb3a0

> Right! Now we're ready to go ahead and download our data. For this tutorial, we'll use the [_LibriSpeech ASR_](https://huggingface.co/datasets/librispeech_asr) corpus. LibriSpeech is one of the most popular datasets for benchmarking speech recognition systems in both academia and industry. It consists of approximately 1000 hours of narrated audiobooks collected from the (LibriVox)[https://librivox.org/] project.

Reviewer: I'm a bit confused here: above we say that we'll use the TEDLIUM dataset, but now we've switched to LibriSpeech. I personally like the TEDLIUM approach more, since LibriSpeech is too common IMO (we want to show something different to every other ASR tutorial :))

Author: I was going to go with TED-LIUM, but there very few models on the Hub trained on this dataset, and by extension no submissions to the leaderboard which is a bit of a shame! In contrast, LibriSpeech is very popular on the Hub and has a very active leaderboard with autoevaluate

> Let’s take a look at the all-time leaderboard on Papers with Code to get a feel for numbers: https://paperswithcode.com/sota/speech-recognition-on-librispeech-test-clean

Reviewer: I suggest using a screenshot here (with link), and also point the reader to our own leaderboards (derived from Hub models): https://huggingface.co/spaces/autoevaluate/leaderboards?dataset=librispeech_asr

We can see that there’s been tremendous progress on the LibriSpeech benchmark in the last three years, with many of the top spots being occupied by Wav2Vec2 or Wav2Vec2-like systems! As mentioned in Section 4, Wav2Vec2 revolutionised the field of speech recognition by introducing an effective pre-training regime, much the same way as BERT revolutionised NLP. We can see that the word error rates (WERs) for this dataset are extremely low! We’ll cover WER in more detail in this section, but for now we can see that the best performing system achieves a WER of just 1.4%, or a word accuracy rate of 100 - 1.4 = 98.6%! This is very impressive! We’ll see how close we can get with our system 🤗

Now that we've got better idea about the dataset that we're working with, let's download it using the `load_dataset()` function:

```python
from datasets import load_dataset, DatasetDict

> raw_datasets = DatasetDict()

Reviewer: Since `datasets` downloads all splits, do we lost a lot of time by simply running?

```
raw_datasets = load_dataset("librispeech_asr", "clean")
```

Author: Switched to streaming mode: https://github.com/huggingface/course/pull/294/commits/629f03bb30d32b93172fe3f138fea25ef9a7d757

The dataset is now downloaded on-the-fly, chunk by chunk

raw_datasets["train"] = load_dataset("librispeech_asr", "clean", split="train.100")
raw_datasets["validation"] = load_dataset("librispeech_asr", "clean", split="validation")
raw_datasets["test"] = load_dataset("librispeech_asr", "clean", split="test")
```

> Note that audio datasets are quite large! For 100h of training data and 10h of validation/test data, we can expect to download ~60GB of raw data.

Reviewer: One constraint to be aware of: the code needs to run on Colab K80s.

Author: Switched to streaming mode!

Author: Running training runs on a K80 now: https://wandb.ai/sanchit-gandhi/hf-course-example
-> should have a compatible training config ready in the next day or so

As it turns out, we represent audio data digitally as a 1-dimensional array. Each array value denotes the amplitude of our audio signal at a particular time step. From the amplitude information, we can reconstruct the frequency spectrum of the audio and recover all the acoustic features.

Since our audio input is continuous, we can't represent an amplitude value for every possible time step. Instead, we _sample_ amplitude values at fixed time steps.

The interval with which we sample our audio is known as the _sampling rate_. For our dataset, we can see that the sampling rate is 16000, meaning 16000 amplitude values are provided each second. Keep the sampling rate in the back of your mind, it'll be important when we come to processing our data later on!

Whilst 1-dimensional arrays are a suitable format for machines, they're not much use to us as humans! We want to be able to **listen** to our audio to get a feel for the speech and recording conditions. Here, we have two options:

1. Convert our 1-dimensional array into a human-friendly format (mp3 or wav).
2. Look to the Hugging Face Hub!

> Option 2 seems much easier to me! Let's head over to the LibriSpeech ASR dataset card on the Hugging Face Hub: https://huggingface.co/datasets/librispeech_asr

Reviewer: I suggest having a screenshot here to guide the reader - great idea with the audio widget!

(Alternatively, we could use a Gradio Space and embed that directly)

    sample = batch["audio"]
    inputs = feature_extractor(sample["array"], sampling_rate=sample["sampling_rate"])
    # process audio length
    batch["input_values"] = inputs.input_values[0]
    batch["input_length"] = len(batch["input_values"])

    # process targets
    input_str = batch["text"].lower()
    # tokenize
    batch["labels"] = tokenizer(input_str).input_ids
    return batch

vectorized_datasets = raw_datasets.map(
    prepare_dataset,
    remove_columns=next(iter(raw_datasets.values())).column_names,
    desc="preprocess dataset",
)
```

Note that it is important to align the decoder's vocabulary with the speech transcriptions of the dataset.

E.g. Librispeech: only captilised letters in the transcriptions, whereas BART was pretrained mostly on lower-cased text.

Filter to max len 20s

```python
MAX_INPUT_LENGTH_IN_SECONDS = 20

max_input_length = MAX_INPUT_LENGTH_IN_SECONDS * feature_extractor.sampling_rate
# filter data that is longer than max_input_length
def is_audio_in_length_range(length):
    return length < max_input_length

vectorized_datasets = vectorized_datasets.filter(
    is_audio_in_length_range,
    input_columns=["input_length"],
)
```

> Define the datacollator:

Reviewer: Do we really need to roll our own collator for this task? Would be nice if this was natively included in `transformers` right?

If that's not possible, please make sure you explain why we need this collator and also remove the docstring to keep the code consise

        # split inputs and labels since they have to be of different lengths and need
        # different padding methods
        input_features = [{"input_values": feature["input_values"]} for feature in features]
        label_features = [{"input_ids": feature["labels"]} for feature in features]

        # pad the audio inputs to max length
        batch = self.processor.feature_extractor.pad(input_features, return_tensors="pt")

        # pad the token targets to max length
        labels_batch = self.processor.tokenizer.pad(label_features, return_tensors="pt")

        # replace padding with -100 to ignore loss correctly
        labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)

        # if bos token is appended in previous tokenization step,
        # cut bos token here as it's appended later
        if (labels[:, 0] == self.decoder_start_token_id).all().cpu().item():
            labels = labels[:, 1:]

        batch["labels"] = labels

        return batch

data_collator = DataCollatorSpeechSeq2SeqWithPadding(
        processor=processor, decoder_start_token_id=model.config.decoder_start_token_id
    )
```

> #### Metrics for speech recognition

Reviewer: FYI here we should use the `evaluate` library if possible
